sync/sync.go

```diff
@@ -322,13 +331,44 @@ func (s *Syncer[H]) requestHeaders(
 
 // storeHeaders updates store with new headers and updates current syncStore's Head.
 func (s *Syncer[H]) storeHeaders(ctx context.Context, headers ...H) error {
+	totalHeaders := len(headers)
+
+	var sbjHeadHeight uint64
+	sbjHead := *s.sbjHead.Load()
+	// if there's no subjective head stored, syncer has already applied the
+	// subjective head to the store and the syncer has passed it
+	if !sbjHead.IsZero() {
+		sbjHeadHeight = sbjHead.Height()
+		// sanity check the height to check whether it's within the range
+		if sbjHeadHeight < headers[0].Height() || sbjHeadHeight > headers[len(headers)-1].Height() {
+			sbjHeadHeight = 0
+		}
+	}
+
+	// if the header is within range, create a segment that *must* be applied
+	// such that the syncer fails if the subjective head cannot be applied
+	if sbjHeadHeight != 0 {
+		segmentSize := sbjHeadHeight - headers[0].Height() + 1
+		mustApplySegment := make([]H, segmentSize)
+		copy(mustApplySegment, headers[:segmentSize])
+		headers = headers[segmentSize:]
+
+		err := s.store.Append(ctx, mustApplySegment...)
+		if err != nil {
+			log.Fatal(fmt.Errorf("failed to apply trusted head at height %d to synced chain segment, "+
+				"potentially following fork: %w", sbjHeadHeight, err))
+		}
+		// if success, then clear subjective head as it's already applied as the store head
+		s.sbjHead.Store(new(H))
+	}
```

Comment:
[suggestion]
Separate this code block with `{` and `}` and add a TODO pointing to backwards sync issue